### ISSUE TITLE
Reward either goes down or stays stagnant

### ISSUE THREAD

**Issue by adhitya-synth** [lvwerra/trl#41]
Let me start off by saying thanks for writing such a wonderful, and easy to use library. I'm genuinely surprised that no one else has created one to approach this technique, considering just how useful it could be. 

I've been trying to adapt trl for my own use case, essentially a summarization task. I've been breaking my head for a few weeks over this. When I try out the positive sentiment generation notebook AS IS, everything seems to work fine. Reward increases over time, KL also slowly approaches the default target value. No sudden peaks or cliffs in any of the loss components.

However, when I try the same with my task, the reward doesn't increase over time. KL also doesn't behave well. They fluctuate VERY wildly, essentially learning nothing. This behaviour slightly changes with different learning rates, but the end result is the same, the output is not what is expected, as you can see in the graph below. What eventually, inevitable happen
s is that at some point, there is a MASSIVE spike in KL divergence, and the reward crashes to 0 (like a bad local minima it struggles to get out of).

<img width="722" alt="image" src="https://user-images.githubusercontent.com/99522885/180948350-b36b16e9-5481-41ff-baab-655b6c9d8fe5.png">

I've tried many variations of the reward function I use, a lot of dataset filtering, but none of it seems to work. I know this isn't a lot of information to work with, but I have another observation that might help you. The original sentiment notebook also fails when the generation input and output lengths are longer. This is quite weird, because it doesn't seem to happen normally with the default length values. I've attached reward and KL mean value graphs for this setup as well (input length: 16->32, output length: 32->64, NO OTHER CODE CHANGES WHATSOEVER). As you can see, no meaningful, consistent reward increase happens, and the KL divergence spikes like crazy (I think once the KL spikes, the model is beyond saving, it just gets stuck in a bad local minima). The graph shown is for the initial part, but I've checked by training it for a long time, and the same thing happens. Could you help me with this, or even just tell me what could be a possible reason? The pink graph is the notebook run as is, no changes, where the graphs and the model behave as expected. Yellow is the same, but with longer input and output sizes.

<img width="726" alt="image" src="https://user-images.githubusercontent.com/99522885/180954578-639cae30-e868-406c-86bc-9a8c9dcca308.png">

**Comment by parshinsh** [comment]
I can confirm that this issue happends. I'm facing the same problem with my own task. Can anyone help with this?

**Comment by Alymostafa** [comment]
same problem here with a longer sequence.
@vblagoje 
@lvwerra

**Comment by Alymostafa** [comment]
@adhitya-synth I used the same configuration as you mentioned and I found out that when the batch size is small it happens as you said but with a larger batch size as in the notebook, the reward increases.

**Comment by hdvvip** [comment]
Recently, I came across OpenAI InstructGPT which is an upgrade version of GPT-3 that has been trained with reinforcement learning.
The reinforcement learning they used for training InstructGPT is PPO which is implemented in this github repository.
Related to the problem that the reward is stagnant or going down, I think even OpenAI (fathers of PPO) also face the same issue. Please see the Figure 13 below.
"As shown in Figure 13, the reward saturates after the initial 400k examples of training."

![Selection_1566](https://user-images.githubusercontent.com/42698038/202613363-c47bc6c4-cc30-45f6-b8de-30d436a6b687.png)

Here is InstructGPT paper.
https://arxiv.org/pdf/2203.02155.pdf

**Comment by hdvvip** [comment]
Thus, based on the OpenAI experiments in InstructGPT paper, I think that it's based on the dataset you used to train your model. In OpenAI case, with the best implementation of PPO, they still failed to improve the rewards when they train GPT-3 using PPO on FLAN and T0 datasets.

![Selection_1567](https://user-images.githubusercontent.com/42698038/202613922-a35816a5-a367-40a6-a6bf-72ca71c04322.png)

**Comment by hdvvip** [comment]
Thus, if you used PPO on your task and it doesn't work. Don't be surprised! Like I said above, some tasks PPO will work. Some tasks, it won't.

**Comment by Alymostafa** [comment]
Thanks for the clarification. But, I am mentioning that based on his observations when the batch size is small what he mentioned happens, but when I increased the batch size I was able to reproduce the same results as in the notebook.

**Comment by hdvvip** [comment]
Well, I think we have some misunderstanding here. I didn't specifically mention you in post. I just want to explain to everyone here that depend on your tasks, PPO may work or not. So, it's not your fault when PPO failed on your NLP task. Everyone here has different tasks, so my answer didn't have anything to do with batch size. BTW, OpenAI used batch size of 128 but still failed.

**Comment by lvwerra** [comment]
Thanks for the discussion here. Indeed, it can depend a lot on the hyperparameters as well as the task. Great you found that increasing the BS works. I think this is still a very underexplored area!

**Comment by leoribeiro** [comment]
@adhitya-synth I face the same problem when using larger text. Did you figure it out a way to overcome this?
